add post on searching for lemmas (#75)
leanprover-community/blog: posts/searching-for-theorems-in-mathlib.md

---
author: 'Bolton Bailey'
category: 'Overview'

> date: 2024-06-01 13:35:44 UTC-08:00

Reviewer: Update before landing

The docs are most useful for users who have some experience in guessing the name of the lemma they are looking for. Helpful in learning how to do this is the [naming conventions page](https://leanprover-community.github.io/contribute/naming.html) which explains how lemmas in mathlib are named (TL;DR theorems are typically named for the declarations they pertain to, with the declaration names appearing in the lemma name in the order they appear in the syntax tree).

A downside of the documentation page is that if a lemma is named in a way you don't expect, it can be hard to find. Also, searching over terms with only a few letters in them can cause the search procedure to hang.

<!-- Note, it would be nice to credit the main developer of the docs page, if there is one. -->

### Moogle

![Moogle Main Page](/images/moogle-screenshot.png)

> [Moogle](https://www.moogle.ai/) is a semantic search engine for Mathlib developed by [Morph labs](https://morph.so/). One enters searches in natural language and the engine tries to turn up theorems that match the search query. For example, if I search for ["the sum of two odd numbers is even"](https://www.moogle.ai/search/raw?q=the%20sum%20of%20two%20odd%20numbers%20is%20even), I turn up the result `Odd.add_odd` which matches this description pretty well as the fourth result.

Reviewer: In my opinion, "semantic search" is not quite an accurate description. Maybe "LLM-based search"?

Author: "Semantic search" is how they described it in their [announcement post](https://leanprover.zulipchat.com/#narrow/channel/113486-announce/topic/Moogle/near/399587531). I will compromise by writing "LLM-based semantic search"

The downside of Moogle is that semantic search isn't an exact science, so the results you are looking for will not always top the result list (as was the case in the example above - a bit of a challenge is that `sum` in mathlib is generally used for indexed sums whereas `add` is more common for the binary operation we were looking for). Moogle can still be a helpful tool, and it doesn't share the weakness of the fuzzy search where it gets confused if the order of the search terms you provide does not exactly match the order of the terms in the lemma name.

### Loogle

![Loogle Main Page](/images/loogle-screenshot.png)

> [Loogle](https://loogle.lean-lang.org/) is a third search engine for mathlib developed by Joachim Breitner at the Lean FRO. It is meant to be a similar tool to [Hoogle](https://hoogle.haskell.org/) from the Haskell community: It allows you to searches for lemmas by their type signature. For example, if I search for [`(?a -> ?b) -> List ?a -> List ?b`](https://loogle.lean-lang.org/?q=(?a%20-%3E%20?b)%20-%3E%20List%20?a%20-%3E%20List%20?b), I get results for `List.map`.

Reviewer 2: ```suggestion
[Loogle](https://loogle.lean-lang.org/) is a third search engine for mathlib developed by Joachim Breitner at the Lean FRO. It is meant to be a similar tool to [Hoogle](https://hoogle.haskell.org/) from the Haskell community: It allows you to search for lemmas by their type signature. For example, if I search for [`(?a -> ?b) -> List ?a -> List ?b`](https://loogle.lean-lang.org/?q=(?a%20-%3E%20?b)%20-%3E%20List%20?a%20-%3E%20List%20?b), I get results for `List.map`.
```

> [Loogle](https://loogle.lean-lang.org/) is a third search engine for mathlib developed by Joachim Breitner at the [Lean Focused Research Organization](https://lean-fro.org/). It is meant to be a similar tool to [Hoogle](https://hoogle.haskell.org/) from the Haskell community: It allows you to searches for lemmas by their type signature. For example, if I search for [`(?a -> ?b) -> List ?a -> List ?b`](https://loogle.lean-lang.org/?q=(?a%20-%3E%20?b)%20-%3E%20List%20?a%20-%3E%20List%20?b), I get results for `List.map`.

Reviewer 2: I mentioned this five months ago...

> [Loogle](https://loogle.lean-lang.org/) is a third search engine for mathlib developed by Joachim Breitner at the [Lean Focused Research Organization](https://lean-fro.org/). It is meant to be a similar tool to [Hoogle](https://hoogle.haskell.org/) from the Haskell community: It allows you to searches for lemmas by their type signature. For example, if I search for [`(?a -> ?b) -> List ?a -> List ?b`](https://loogle.lean-lang.org/?q=(?a%20-%3E%20?b)%20-%3E%20List%20?a%20-%3E%20List%20?b), I get results for `List.map`.

Author: Weird it says "suggestion applied" I wonder why it didn't go through.

> [Loogle](https://loogle.lean-lang.org/) is a third search engine for mathlib developed by Joachim Breitner at the [Lean Focused Research Organization](https://lean-fro.org/). It is meant to be a similar tool to [Hoogle](https://hoogle.haskell.org/) from the Haskell community: It allows you to searches for lemmas by their type signature. For example, if I search for [`(?a -> ?b) -> List ?a -> List ?b`](https://loogle.lean-lang.org/?q=(?a%20-%3E%20?b)%20-%3E%20List%20?a%20-%3E%20List%20?b), I get results for `List.map`.

Author: Maybe I clicked the wrong button, sorry about that

> [Loogle](https://loogle.lean-lang.org/) is a third search engine for mathlib developed by Joachim Breitner at the [Lean Focused Research Organization](https://lean-fro.org/). It is meant to be a similar tool to [Hoogle](https://hoogle.haskell.org/) from the Haskell community: It allows you to searches for lemmas by their type signature. For example, if I search for [`(?a -> ?b) -> List ?a -> List ?b`](https://loogle.lean-lang.org/?q=(?a%20-%3E%20?b)%20-%3E%20List%20?a%20-%3E%20List%20?b), I get results for `List.map`.

Author: ```suggestion
[Loogle](https://loogle.lean-lang.org/) is a third search engine for mathlib developed by Joachim Breitner at the [Lean Focused Research Organization](https://lean-fro.org/). It is meant to be a similar tool to [Hoogle](https://hoogle.haskell.org/) from the Haskell community: It allows you to search for lemmas by their type signature. For example, if I search for [`(?a -> ?b) -> List ?a -> List ?b`](https://loogle.lean-lang.org/?q=(?a%20-%3E%20?b)%20-%3E%20List%20?a%20-%3E%20List%20?b), I get results for `List.map`.
```

> Loogle is a great tool for finding lemmas that you know the type of, but not the name of. It is a pretty precise tool; a downside is that it will not by default give you responses without you typing out the names of the definitions relevant to your search correctly and in full. To avoid putting in too many restrictions and getting an empty result list, it's recommended that you add search filters one at a time to your query.

Reviewer: I personally tend to use it more with just a list of definitions that are used rather than a full type signature. Maybe "...by their type signature or types that occur in it"

## Other websites

There are a few other websites that can be helpful for finding theorems in mathlib, although they are not conventional search engines like the ones above.

> 

Reviewer 3: I think that leansearch.net should be mentioned.

Author: I think the reason it is not already present is that leansearch was released after I wrote this. Now, six months later, I am seeing that there is another tool, lean explore. Perhaps I should rework this section to be about LLM-based search engines in general.

### Zulip

> The [Zulip chat](https://leanprover.zulipchat.com/) for the Lean community has a channel called ["Is there code for X?"](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F) where you can ask questions about the presence of particular theorems in mathlib, major and minor. If you are having trouble finding a theorem, you can ask in this channel and someone will likely be able to help you, although it might take longer than the search engine would.

Reviewer: Perhaps add "You can also search the archives to see if someone has asked about the same results before."

* The [Library overview](https://leanprover-community.github.io/mathlib-overview.html): This page lists the overall most important high-level declarations in mathlib, organized by topic/subfield of mathematics.
* The [Undergrad list](https://leanprover-community.github.io/undergrad/) is a list of theorems and declarations in mathlib that are considered to be at an undergraduate level. This list comes from a document compiled by the [French Ministry of Education](https://media.devenirenseignant.gouv.fr/file/agreg_externe/59/7/p2020_agreg_ext_maths_1107597.pdf)
* [Wiedijk's 100 Theorems](https://leanprover-community.github.io/100.html): This page lists 100 important theorems from a variety of fields of mathematics. They were originally compiled by Freek Wiedijk in his [Formalizing 100 Theorems](https://www.cs.ru.nl/~freek/100/) project as a benchmark for proof assistants. The page has links to the mathlib documentation page for each theorem that mathlib has proven, as well as a link to a page with the list of theorems that mathlib has not (yet!) proven.

## Tactics

> In addition to the search engines and webpages, there are a few tactics you can use to find theorems in mathlib without leaving your lean development environment.

Author: ```suggestion
In addition to the search engines and webpages, there are a few tactics you can use to find theorems in mathlib without leaving your Lean development environment.
```

> In addition to the search engines and webpages, there are a few tactics you can use to find theorems in mathlib without leaving your lean development environment.

Author: lean -> Lean

### `exact?`

> The `exact?` tactic (formerly known as `library_search`) is a great way to find theorems in mathlib that finish off a proof goal. At any point in a tactic proof, you can type `exact?` and lean will search through the lemmas you have imported (as well as the file you are working in itself) for a theorem that proves the current goal using assumption from you context. If it finds one, it will show a helpful message saying `Try this:` which you can click to insert the theorem into your proof.

Author: assumption -> assumptions

Author: ```suggestion
The `exact?` tactic (formerly known as `library_search`) is a great way to find theorems in mathlib that finish off a proof goal. At any point in a tactic proof, you can type `exact?` and lean will search through the lemmas you have imported (as well as the file you are working in itself) for a theorem that proves the current goal using assumptions from you context. If it finds one, it will show a helpful message saying `Try this:` which you can click to insert the theorem into your proof.
```

## Other websites

There are a few other websites that can be helpful for finding theorems in mathlib, although they are not conventional search engines like the ones above.

### Zulip

The [Zulip chat](https://leanprover.zulipchat.com/) for the Lean community has a channel called ["Is there code for X?"](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F) where you can ask questions about the presence of particular theorems in mathlib, major and minor. If you are having trouble finding a theorem, you can ask in this channel and someone will likely be able to help you, although it might take longer than the search engine would. You can also search the archives to see if someone has asked about the same results before.

### The Library overview/The Undergrad list/Wiedijk's 100 Theorems

> The sidebar of [main page](https://leanprover-community.github.io/index.html) for the mathlib community lists three additional pages under the "Library Overiviews" heading. These pages have links to the docs for fixed collections of major declarations in mathlib. These are:

Reviewer 4: Typo 

```suggestion
The sidebar of [main page](https://leanprover-community.github.io/index.html) for the mathlib community lists three additional pages under the "Library Overviews" heading. These pages have links to the docs for fixed collections of major declarations in mathlib. These are:
```

## Tactics

In addition to the search engines and webpages, there are a few tactics you can use to find theorems in mathlib without leaving your Lean development environment.

### `exact?`

The `exact?` tactic (formerly known as `library_search`) is a great way to find theorems in mathlib that finish off a proof goal. At any point in a tactic proof, you can type `exact?` and lean will search through the lemmas you have imported (as well as the file you are working in itself) for a theorem that proves the current goal using assumptions from your context. If it finds one, it will show a helpful message saying `Try this:` which you can click to insert the theorem into your proof.

A good time to use `exact?` is when you have a proof goal that you think is so straightforward that it is likely to be in mathlib already. If `exact?` doesn't find a proof, you can always extract the fact and prove it or `sorry` it as an independent lemma immediately above the theorem you are working on, and `exact?` should start to work. You can then prove the goal yourself and submit the proof to mathlib.

> A downside of this approach is that it depnds on the *exact* structure of the goal you are trying to prove being present in the library. If you are hoping to prove the goal `a - b = c` and the library only has a form of the theorem that looks like `a = b + c`, `exact?` will not find it. (Helpful in this process is experience with how mathlib prefers theorems to be structured. For example, a form of a theorem with a subtraction is discouraged in favor of a form with an addition, and `>` is almost entirely absent in favor of `<`.)

Reviewer 4: Typo

```suggestion
A downside of this approach is that it depends on the *exact* structure of the goal you are trying to prove being present in the library. If you are hoping to prove the goal `a - b = c` and the library only has a form of the theorem that looks like `a = b + c`, `exact?` will not find it. (Helpful in this process is experience with how mathlib prefers theorems to be structured. For example, a form of a theorem with a subtraction is discouraged in favor of a form with an addition, and `>` is almost entirely absent in favor of `<`.)
```

### `rw?`

In the same vein as `exact?` and `apply?`, `rw?` is a tactic that searches for lemmas that can be used to rewrite the current goal. It's useful when you have a subterm of your goal that you know can be rewritten into another form, but you don't know the name of the lemma that does the rewriting.

### Other Automated Theorem Proving Techniques

There are a few other tactics in mathlib and other projects that don't search for theorems per se, but which try to complete proofs automatically, and which may identify useful theorems in the process. These include:

* `simp?`: Simplifies the goal using a set of simplification rules.
* `aesop`: From the [Aesop project](https://github.com/leanprover-community/aesop). A tactic that tries to prove the current goal using a search tree.
* `rw_search`: Attempts to find chains of rewrites that will prove or simplify the current goal.
* `hint`: A "Kitchen Sink" tactic that attempts to close the goal using a variety of other tactics.

> 

Reviewer 4: You might want to include [`canonical`](https://github.com/chasenorman/CanonicalLean) here too.

And ones that are not imported by mathlib:

> 

Reviewer 4: You might want to include [`canonical`](https://github.com/chasenorman/CanonicalLean) here too.

Author: Added!

---
author: 'Bolton Bailey'
category: 'Overview'
date: 2025-05-15 1:26:31 UTC-08:00
description: ''
has_math: false
link: ''
slug: searching-for-theorems-in-mathlib
tags: ''
title: Searching for Theorems in Mathlib
type: text
---
A post for beginners on the different ways to search for theorems in mathlib, inspired by [this talk](https://www.youtube.com/watch?v=UJrYKR01QwU).

<!-- TEASER_END -->

Mathlib has a lot of theorems, and it can be hard to find the one you're looking for. This post goes over various search engines, webpages and tactics you can use to find theorems in mathlib. For each of them, I'll give a brief description of how they work and what their strengths and weaknesses are.

## Search engines

There are three main search engines which allow you to search for theorems in mathlib and its dependencies.

### The mathlib4 docs page

![Docs Main Page](/images/documentation-screenshot.png)

> The most comprehensive listing of mathlib theorems is the [documentation page on the community website](https://leanprover-community.github.io/mathlib4_docs/). It's a documentation page just as you might find for any other software project: On the left it has nested dropdown menus for the directory structure of mathlib and several of its dependencies. In the upper right, it has a search bar. The search bar provides a "fuzzy" search which searches for theorem names that contain the search term as a substring. For example, if I search for ["**foobar**"](https://leanprover-community.github.io/mathlib4_docs/search.html?sitesearch=https%3A%2F%2Fleanprover-community.github.io%2Fmathlib4_docs&q=foobar), the current top result is [**F**l**o**at.**o**f**B**in**ar**yScientific](https://leanprover-community.github.io/mathlib4_docs/Init/Data/OfScientific.html#Float.ofBinaryScientific).

Author: I think it's right to capitalize a sentence after a color. I will add a comma though.

The docs are most useful for users who have some experience in guessing the name of the lemma they are looking for. Helpful in learning how to do this is the [naming conventions page](https://leanprover-community.github.io/contribute/naming.html) which explains how lemmas in mathlib are named (TL;DR theorems are typically named for the declarations they pertain to, with the declaration names appearing in the lemma name in the order they appear in the syntax tree).

> A downside of the documentation page is that if a lemma is named in a way you don't expect, it can be hard to find. Also, searching over terms with only a few letters in them can cause the search procedure to hang.

Author: I guess this PR changed it https://github.com/leanprover/doc-gen4/pull/293

> A downside of the documentation page is that if a lemma is named in a way you don't expect, it can be hard to find. Also, searching over terms with only a few letters in them can cause the search procedure to hang.

Author: I have deleted this line

> <!-- Note, it would be nice to credit the main developer of the docs page, if there is one. -->

Author: I will mention the main developers and refer the reader to the source to see the full list of contributors.

> ### Moogle and other LLM-based Search tools

Author: Ok, I've reworked the section to not mention moogle specifically in the main text, though I'm still using the screenshot as the example.

I will still put a bulleted list of the tools that exist today at the bottom of the section. The point of this post is to direct people to solutions, and I think that goal is better served if we include links.

![Moogle Main Page](/images/moogle-screenshot.png)

> [Moogle](https://www.moogle.ai/) is an LLM-based semantic search engine for Mathlib developed by [Morph labs](https://morph.so/). One enters searches in natural language and the engine tries to turn up theorems that match the search query. For example, if I search for ["the sum of two odd numbers is even"](https://www.moogle.ai/search/raw?q=the%20sum%20of%20two%20odd%20numbers%20is%20even), I turn up the result `Odd.add_odd` which matches this description pretty well as the fourth result.

Author: No, I think semantic search is the right terminology - this is how these projects refer to what they do, i.e. using NLP tools like embeddings to search. Fuzzy search refers to search that accommodates misspellings or partial spellings, like the VSCode file search bar.

I thought I had already pushed a change to turn this into a section that mentioned moogle, leansearch and leanexplore, but coming back to this, I don't see those changes. I'll try to rewrite it again.

> [Moogle](https://www.moogle.ai/) is an LLM-based semantic search engine for Mathlib developed by [Morph labs](https://morph.so/). One enters searches in natural language and the engine tries to turn up theorems that match the search query. For example, if I search for ["the sum of two odd numbers is even"](https://www.moogle.ai/search/raw?q=the%20sum%20of%20two%20odd%20numbers%20is%20even), I turn up the result `Odd.add_odd` which matches this description pretty well as the fourth result.

Author: Ok sorry I see I actually had changed this section to include two bullet points at the end. Perhaps the issue is that I should not mention "Moogle" since it's so outdated?

> The downside of Moogle is that semantic search isn't an exact science, so the results you are looking for will not always top the result list (as was the case in the example above - a bit of a challenge is that `sum` in mathlib is generally used for indexed sums whereas `add` is more common for the binary operation we were looking for). Moogle can still be a helpful tool, and it doesn't share the weakness of the fuzzy search where it gets confused if the order of the search terms you provide does not exactly match the order of the terms in the lemma name.

Author: Again, this is semantic search, here is a screenshot showing that the title of the moogle search page describes it as semantic search, and links to the arxiv papers for leansearch.net and leanexplore.net, both of which call what they do semantic search in their title or abstract.

<img width="381" alt="Screenshot 2025-06-19 at 8 59 03 AM" src="https://github.com/user-attachments/assets/f7a64dd8-0f4c-488f-a034-a7aa57966854" />

https://arxiv.org/abs/2403.13310v2
https://arxiv.org/html/2506.11085v1

Since the release of Moogle, other LLM-based search engines have come out, offering other features:

* [LeanSearch](https://leansearch.net/) from the BICMR AI for Mathematics team at PKU, which offers the ability to augment a previous query.
* [LeanExplore](https://www.leanexplore.com/) from Justin Asher, which offers the ability to select different other Lean libraries to query.

These can be good alternatives to Moogle, depending on what you prefer in your search experience.

### Loogle

![Loogle Main Page](/images/loogle-screenshot.png)

> [Loogle](https://loogle.lean-lang.org/) is a third search engine for mathlib developed by Joachim Breitner at the [Lean Focused Research Organization](https://lean-fro.org/). It is meant to be a similar tool to [Hoogle](https://hoogle.haskell.org/) from the Haskell community: It allows you to search for lemmas by their type signature, or by listing types or subexpressions that should occur in the lemma. For example, if I search for [`(?a -> ?b) -> List ?a -> List ?b`](https://loogle.lean-lang.org/?q=(?a%20-%3E%20?b)%20-%3E%20List%20?a%20-%3E%20List%20?b), I get results for `List.map`. 

Author: I was taught that it is right to put a capital after a colon when  it's a complete sentence. Is there some kind of style manual that has been mandated for this blog I should adhere to that says otherwise?

> [Loogle](https://loogle.lean-lang.org/) is a third search engine for mathlib developed by Joachim Breitner at the [Lean Focused Research Organization](https://lean-fro.org/). It is meant to be a similar tool to [Hoogle](https://hoogle.haskell.org/) from the Haskell community: It allows you to search for lemmas by their type signature, or by listing types or subexpressions that should occur in the lemma. For example, if I search for [`(?a -> ?b) -> List ?a -> List ?b`](https://loogle.lean-lang.org/?q=(?a%20-%3E%20?b)%20-%3E%20List%20?a%20-%3E%20List%20?b), I get results for `List.map`. 

Author: I guess this is an APA vs Chicago difference? I guess it doesn't really matter, so I'll just change it.

> Loogle is a great tool for finding lemmas that you know the type of, but not the name of. It is a pretty precise tool; a downside is that it will not by default give you responses without you typing out the names of the definitions relevant to your search correctly and in full. To avoid putting in too many restrictions and getting an empty result list, it's recommended that you add search filters one at a time to your query.

Reviewer 5: Loogle has lots of useful little features: it's probably not necessary to explain in full, but one could hint at this.

> Loogle is a great tool for finding lemmas that you know the type of, but not the name of. It is a pretty precise tool; a downside is that it will not by default give you responses without you typing out the names of the definitions relevant to your search correctly and in full. To avoid putting in too many restrictions and getting an empty result list, it's recommended that you add search filters one at a time to your query.

Author: I'll go over this section to try to make these facts clear.

### The Library overview/The Undergrad list/Wiedijk's 100 Theorems

The sidebar of [main page](https://leanprover-community.github.io/index.html) for the mathlib community lists three additional pages under the "Library Overviews" heading. These pages have links to the docs for fixed collections of major declarations in mathlib. These are:

* The [Library overview](https://leanprover-community.github.io/mathlib-overview.html): This page lists the overall most important high-level declarations in mathlib, organized by topic/subfield of mathematics.
* The [Undergrad list](https://leanprover-community.github.io/undergrad/) is a list of theorems and declarations in mathlib that are considered to be at an undergraduate level. This list comes from a document compiled by the [French Ministry of Education](https://media.devenirenseignant.gouv.fr/file/agreg_externe/59/7/p2020_agreg_ext_maths_1107597.pdf)

> * [Wiedijk's 100 Theorems](https://leanprover-community.github.io/100.html): This page lists 100 important theorems from a variety of fields of mathematics. They are tracked by Freek Wiedijk in his [Formalizing 100 Theorems](https://www.cs.ru.nl/~freek/100/) project as a benchmark for proof assistants. The page has links to the mathlib documentation page for each theorem that mathlib has proven, as well as a link to a page with the list of theorems that mathlib has not (yet!) proven.

Reviewer 5: There's also the 1000+ theorems lists, which lists formalisation entries for (roughly) all theorems which have a named wikipedia page. (This is not fully accurate, but that's a bug in wikipedia/the theorem data collection.)

* `simp?`: Simplifies the goal using a set of simplification rules.
* `rw_search`: Attempts to find chains of rewrites that will prove or simplify the current goal.
* `hint`: A "Kitchen Sink" tactic that attempts to close the goal using a variety of other tactics.
* [`aesop`](https://github.com/leanprover-community/aesop): A tactic that tries to prove the current goal using a search tree.

And ones that are not imported by mathlib:

* `auto`: From the [Lean Auto project](https://github.com/leanprover-community/lean-auto). A tactic that tries to prove the current goal using a technique "based on a monomorphization procedure from dependent type theory to higher-order logic and a deep embedding of higher-order logic into dependent type theory".
* `suggest_tactics`/`search_proof`: From the [LeanCopilot project](https://github.com/lean-dojo/LeanCopilot). A tactic that tries to suggest tactics using an LLM.
* `llmstep`: From the [Llemma project](https://github.com/wellecks/llmstep). Another tactic that tries to suggest tactics using an LLM.

> * `canonical`: From [Chase Norman](https://github.com/chasenorman/CanonicalLean). An exhaustive search procedure for terms in dependent type theory. 

Reviewer 5: (And `grind` in the future --- thought it should probably not be added yet.)

---
author: 'Bolton Bailey'
category: 'Overview'
date: 2025-05-15 1:26:31 UTC-08:00
description: ''
has_math: false
link: ''
slug: searching-for-theorems-in-mathlib
tags: ''
title: Searching for Theorems in Mathlib
type: text
---

> A post for beginners on the different ways to search for theorems in mathlib, inspired by [this talk](https://www.youtube.com/watch?v=UJrYKR01QwU)  from Jireh Loreaux.

Reviewer 6: ```suggestion
A post for beginners on the different ways to search for theorems in mathlib, inspired by [this talk](https://www.youtube.com/watch?v=UJrYKR01QwU) from Jireh Loreaux.
```

<!-- TEASER_END -->

Mathlib has a lot of theorems, and it can be hard to find the one you're looking for. This post goes over various search engines, webpages and tactics you can use to find theorems in mathlib. For each of them, I'll give a brief description of how they work and what their strengths and weaknesses are.

## Search engines

> There are three main search engines which allow you to search for theorems in mathlib and its dependencies.

Reviewer 3: ```suggestion
There are three types of search engine which allow you to search for theorems in mathlib and its dependencies.
```

### The mathlib4 docs page

![Docs Main Page](/images/documentation-screenshot.png)

> The most comprehensive listing of mathlib theorems is the [documentation page on the community website](https://leanprover-community.github.io/mathlib4_docs/). This page was developed for Lean 3 by Rob Lewis and for Lean 4 by Henrik Böving and others, see the [docGen source for a list of contributors](https://github.com/leanprover/doc-gen4). It's a documentation page just as you might find for any other software project: on the left, it has nested dropdown menus for the directory structure of mathlib and several of its dependencies. In the upper right, it has a search bar. The search bar provides a "fuzzy" search which searches for theorem names that contain the search term as a substring. For example, if I search for ["**foobar**"](https://leanprover-community.github.io/mathlib4_docs/search.html?sitesearch=https%3A%2F%2Fleanprover-community.github.io%2Fmathlib4_docs&q=foobar), the current top result is [**F**GM**o**duleCat.**ob**j_c**ar**rier](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Category/FGModuleCat/Basic.html#FGModuleCat.obj_carrier).

Reviewer 6: ```suggestion
The most comprehensive listing of mathlib theorems is the [documentation page on the community website](https://leanprover-community.github.io/mathlib4_docs/). This page was developed for Lean 3 by Rob Lewis and for Lean 4 by Henrik Böving and others, see the [docGen source](https://github.com/leanprover/doc-gen4) for a list of contributors. It's a documentation page just as you might find for any other software project: on the left, it has nested dropdown menus for the directory structure of mathlib and several of its dependencies. In the upper right, it has a search bar. The search bar provides a "fuzzy" search which searches for theorem names that contain the search term as a substring. For example, if I search for ["**foobar**"](https://leanprover-community.github.io/mathlib4_docs/search.html?sitesearch=https%3A%2F%2Fleanprover-community.github.io%2Fmathlib4_docs&q=foobar), the current top result is [**F**GM**o**duleCat.**ob**j_c**ar**rier](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Category/FGModuleCat/Basic.html#FGModuleCat.obj_carrier).
```

Reviewer 6: Also, maybe add a line break or two for legibility of the source?

The docs are most useful for users who have some experience in guessing the name of the lemma they are looking for. Helpful in learning how to do this is the [naming conventions page](https://leanprover-community.github.io/contribute/naming.html) which explains how lemmas in mathlib are named (in short, theorems are typically named for the declarations they pertain to, with the declaration names appearing in the lemma name in the order they appear in the syntax tree).

A downside of the documentation page is that if a lemma is named in a way you don't expect, it can be hard to find.

> 

Reviewer 7: I would mention tab completion somewhere, either here because it's the same kind of fuzzy algorithm as in the docs page, or when you talk about tactics because it's something you do while doing some Lean.

The downside of this approach is that semantic search isn't an exact science, so the results you are looking for will not always top the result list (in case of the example above - a bit of a challenge is that `sum` in mathlib is generally used for indexed sums whereas `add` is more common for the binary operation we were looking for). These can still be helpful tools, as they don't share the weakness of the fuzzy search where it gets confused if the order of the search terms you provide does not exactly match the order of the terms in the lemma name.

A number of NLP-based search engines have come out, offering different features:

* [Moogle](https://www.moogle.ai/) is an LLM-based semantic search engine for Mathlib developed by [Morph labs](https://morph.so/). This was one of the first search engines of this type, though at this point it's somewhat outdated.
* [LeanSearch](https://leansearch.net/) from the BICMR AI for Mathematics team at PKU, which offers the ability to augment a previous query.

> * [LeanExplore](https://www.leanexplore.com/) from Justin Asher, which offers the ability to select different other Lean libraries to query.

Reviewer 6: ```suggestion
* [LeanExplore](https://www.leanexplore.com/) from Justin Asher, which offers the ability to select different Lean libraries to query.
```
maybe?

### `apply?`

The `apply?` tactic is similar to `exact?`. It searches for theorems that can be applied to the current goal, without necessarily requiring that they close the goal (although if they do close the goal, `apply?` will show you that).

`apply?` is useful when the structure of your goal is such that there seems to be an obvious fact to apply, but you don't know the name of the fact. It's pretty much strictly more powerful than `exact?`, but it can be slower to run.

### `rw?`

In the same vein as `exact?` and `apply?`, `rw?` is a tactic that searches for lemmas that can be used to rewrite the current goal. It's useful when you have a subterm of your goal that you know can be rewritten into another form, but you don't know the name of the lemma that does the rewriting. You can also use this to rewrite hypotheses with `rw? at ...`.

### Other Automated Theorem Proving Techniques

There are a few other tactics in mathlib and other projects that don't search for theorems per se, but which try to complete proofs automatically, and which may identify useful theorems in the process. These include:

> * `simp`: Simplifies the goal using a set of simplification rules. Using `simp?` allows you to replace `simp` with an equivalent tactic that specifies the simplification rules it uses.

Reviewer 7: Can you mention that `simp?` is a good way to learn the lemmas name, that you may then use in a different way if you like (`rw`, proof term, etc.)?

---
author: 'Bolton Bailey'
category: 'Overview'
date: 2025-05-15 1:26:31 UTC-08:00
description: ''
has_math: false
link: ''
slug: searching-for-theorems-in-mathlib
tags: ''
title: Searching for Theorems in Mathlib
type: text
---
A post for beginners on the different ways to search for theorems in mathlib, inspired by [this talk](https://www.youtube.com/watch?v=UJrYKR01QwU) from Jireh Loreaux.

<!-- TEASER_END -->

Mathlib has a lot of theorems, and it can be hard to find the one you're looking for. 
This post goes over various search engines, webpages and tactics you can use to find theorems in mathlib. 
For each of them, I'll give a brief description of how they work and what their strengths and weaknesses are.

> ## Search engines

Reviewer 8: ```suggestion
# Search engines
```
Nikola automatically increments headers one level.

There are three types of search engine which allow you to search for theorems in mathlib and its dependencies.

> ### The mathlib4 docs page

Reviewer 8: ```suggestion
## The mathlib4 docs page
```

the current top result is [**F**GM**o**duleCat.**ob**j_c**ar**rier](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Category/FGModuleCat/Basic.html#FGModuleCat.obj_carrier). 
This is therefore a similar process to searching for a declaration using an in-editor code completion tool such as [IntelliSense](https://code.visualstudio.com/docs/editing/intellisense)
(which is itself a useful way to search for lemmas while coding in Lean).

The docs are most useful for users who have some experience in guessing the name of the lemma they are looking for. 
Helpful in learning how to do this is the [naming conventions page](https://leanprover-community.github.io/contribute/naming.html) 
which explains how lemmas in mathlib are named 
(in short, theorems are typically named for the declarations they pertain to, 
with the declaration names appearing in the lemma name in the order they appear in the syntax tree).

A downside of the documentation page is that if a lemma is named in a way you don't expect, 
it can be hard to find.

<!-- Note, it would be nice to credit the main developer of the docs page, if there is one. -->

> ### Natural language search tools

Reviewer 8: ```suggestion
## Natural language search tools
```

For example, if I search for [`(?a -> ?b) -> List ?a -> List ?b`](https://loogle.lean-lang.org/?q=(?a%20-%3E%20?b)%20-%3E%20List%20?a%20-%3E%20List%20?b), 
I get results for `List.map`. 
If I search for [`Real.sin, Real.cos, Real.tan`](https://loogle.lean-lang.org/?q=Real.sin%2C+Real.cos%2C+Real.tan), 
I get results for the theorem that `Real.tan x = Real.sin x / Real.cos x`.

Loogle is a great tool for finding lemmas when you know the types used, but not the name of the lemma itself. 
It is a pretty precise tool; a downside is that it will not by default give you specific responses without you typing out the names of at least 2 definitions relevant to your search correctly and in full 
(though you can revert to substring-based search using quotes, see the page for a list of search techniques you can use). 
To avoid putting in too many restrictions and getting an empty result list, 
it's recommended that you add search filters one at a time to your query.
This search engine is also available in the Lean4 VSCode extension,
via the "Lean4: Loogle: Search" command.

> ## Other websites

Reviewer 8: ```suggestion
# Other websites
```

There are a few other websites that can be helpful for finding theorems in mathlib, 
although they are not conventional search engines like the ones above.

> ### Zulip

Reviewer 8: ```suggestion
## Zulip
```

The [Zulip chat](https://leanprover.zulipchat.com/) for the Lean community has a channel called ["Is there code for X?"](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F) 
where you can ask questions about the presence of particular theorems in mathlib, major and minor.
If you are having trouble finding a theorem, 
you can ask in this channel and someone will likely be able to help you, 
although it might take longer than the search engine would. 
You can also search the archives to see if someone has asked about the same results before.

> ### Lists of specific theorems

Reviewer 8: ```suggestion
## Lists of specific theorems
```
